R/tbl_split.R

```diff
@@ -1,66 +1,257 @@
+#' @param footnotes (`string`)\cr
+#'   can be either `"first"`, `"each"`, or `"last"`, to locate global footnotes
+#'   only on the first, in each, or in the last table, respectively. It defaults
+#'   to `"each"`. Reference footnotes are always present wherever they appear.
+#' @param caption (`string`)\cr
+#'   similarly to footnotes, makes the titles, and subtitles appear only on the
+#'   `"first"` or `"each"` table. It defaults to `"each"`.
 #' @inheritParams rlang::args_dots_empty
 #'
-#' @return `tbl_split` object
-#' @name tbl_split
+#' @return `tbl_split` object. They can also be identified further by subclass
+#'   definitions `tbl_split_by_rows` and `tbl_split_by_columns`.
+#'
+#' @details
+#' Run [show_header_names()] to print all column names to split by.
 #'
 #' @examples
-#' tbl <-
-#'   tbl_summary(trial) |>
-#'   tbl_split(variables = c(marker, grade))
+#' # create standard table
+#' tbl <- tbl_summary(trial, by = trt)
+#'
+#' # split by rows ---------------------------
+#' tbl_list_rows <- tbl |>
+#'   tbl_split_by_rows(variables = c(marker, grade))
+#'
+#' # print column names
+#' show_header_names(tbl)
+#'
+#' # split by columns ------------------------
+#' tbl_list_cols <- tbl |>
+#'   tbl_split_by_columns(groups = list("stat_1", "stat_2"))
+#'
+#' # take the list of splitted rows and split again by columns
+#' tbl_list_rows_cols <- tbl_list_rows |>
+#'   tbl_split_by_columns(groups = list("stat_1", "stat_2"))
+#'
+#' # Handle footnotes -------------------------
+#' tbl_fc <- trial |>
+#' tbl_summary(by = trt, missing = "no") |>
+#'   modify_footnote_header(
+#'     footnote = "All but four subjects received both treatments in a crossover design",
+#'     columns = all_stat_cols(),
+#'     replace = FALSE
+#'   ) |>
+#'   modify_footnote_body(
+#'     footnote = "Tumor grade was assessed _before_ treatment began",
+#'     columns = "label",
+#'     rows = variable == "grade" & row_type == "label"
+#'   ) |>
+#'   modify_spanning_header(
+#'     c(stat_1, stat_2) ~ "**TRT**"
+#'   ) |>
+#'   modify_abbreviation("I = 1, II = 2, III = 3") |>
+#'   modify_caption("_Some caption_") |>
+#'   modify_footnote_spanning_header(
+#'     footnote = "Treatment",
+#'     columns = c(stat_1)
+#'   ) |>
+#'   modify_source_note("Some source note!")
+#'
+#' # lets split it!
+#' tbl_list <- tbl_fc |>
+#'   tbl_split_by_rows(variables = c(marker, grade), footnotes = "last", caption = "first")
+#'
+#' @name tbl_split_by
 NULL
 
 #' @export
-#' @rdname tbl_split
-tbl_split <- function(x, ...) {
-  UseMethod("tbl_split")
+#' @rdname tbl_split_by
+tbl_split_by_rows <- function(x, variables, footnotes = "each", caption = "each") {
+  set_cli_abort_call()
+
+  # list map -------------------------------------------------------------------
+  if (inherits(x, "list")) {
+    tbl_list <- map(
+      .x = x,
+      .f = tbl_split_by_rows,
+      variables = {{ variables }}
+    )
+  } else {
```

Comment:
Instead of wrapping all the code in large `if () else ()` chunks, can you update the first chunk to 

```r
if (inherits(x, "tbl_split")) {
  return(...)
}
```

then we don't need the big `else ()` chunk. Same applies to the column split function

Reply:
It was like this but I changed to process footnotes only once. If you want only one if() I need to duplicate that handling